openlibrary/coverstore/warc.py

```diff
@@ -284,10 +286,10 @@ def write(self, record):
         """Writes a record into the WARC file.
         Assumes that data_length and other attributes are correctly set in record.header.
         """
-        self.file.write(str(record.get_header()).encode('utf-8'))
+        self.file.write(str(record.get_header()))
```

Comment:
In a Python 3-esque world, I'd prefer to see this be `bytes(record.get_header())`. Not at all a showstopper, but it would make me relax a little.

Reply:
Let's discuss this sometime.  Perhaps I have been thinking about this wrong.

Reply:
Happy to! I'll type out some thoughts here, and we can discuss more on Slack or elsewhere.

As context, when I typed it, I didn't know what `record.get_header()` is or what it returns. One key question for me, in a Python 3 world, is: is `record.get_header()` text that might get shown to a user, or is it instead bytes that might get written to/from disk?

I asked myself the same question about `WARCHeader` in general. From what I can tell, `WARC` is first and foremost a file format.

To me, that means that `self.file` should be a byte (`BytesIO`) file, not a `StringIO` file. (For justification of this, see e.g. https://github.com/webrecorder/warcio/issues/7 -- it's technically possible for the WARC file to contain non-UTF-8 data. In that case, it's helpful for our layer IMHO to be able to handle such a situation.)

Since I propose that `self.file` is going to stay `bytes`, then we'd need to make sure `bytes` get passed through into it. That's why I wrote `bytes(record.get_header())` in the comment. Now that I think about it, it's probably better if `record.get_header()` just **is** bytes, and then we wouldn't call `bytes()` on it.

I hope that helps. It's not necessarily the most clear explanation, so I'm happy to discuss more. It's also just one way of thinking about the situation, and I'm new to this codebase.